boot/freeldr/freeldr/arch/uefi/i386/uefiasm.S

```diff
@@ -0,0 +1,119 @@
+/*
+ * PROJECT:     FreeLoader UEFI Support
+ * LICENSE:     GPL-2.0-or-later (https://spdx.org/licenses/GPL-2.0-or-later)
+ * PURPOSE:     i386 assembly UEFI escape helper functions
+ * COPYRIGHT:   Copyright 2023 Justin Miller <justinmiller100@gmail.com>
+ */
+
+#include <asm.inc>
+#include <arch/pc/x86common.h>
+#include <arch/pc/pcbios.h>
+
+PUBLIC _gdtptr
+PUBLIC _i386idtptr
+PUBLIC __changestack
+PUBLIC __changegdt
+EXTERN _UefiServiceStack:DWORD
+EXTERN _i386Idt:DWORD
+EXTERN _ExecuteLoaderCleanly:PROC
+EXTERN _UefiExitBootServices:PROC
+EXTERN _BasicStack:DWORD
+.code32
+
+// void __exituefi(VOID)
+PUBLIC __exituefi
+__exituefi:
+    push ebp
+    push esi
+    push edi
+    push ebx
+
+    /* Save the old stack */
+    mov ebx, esp
+
+    /* Load the new stack */
+    xor ebp, ebp
+    mov esp, _UefiServiceStack
+
+    /* Call the entry routine, passing the parameters */
+    call _UefiExitBootServices
+
+    /* Retore old stack */
+    mov esp, ebx
+
+    /* Retore non-volatiles */
+    pop ebx
+    pop edi
+    pop esi
+    pop ebp
+
+
+    /* All done */
+    ret
+
+// void __reloadsegment(VOID)
+PUBLIC __changestack
+__changestack:
+    mov eax, esp
+    mov esp, _BasicStack
+    push eax
+    call _ExecuteLoaderCleanly
```

Comment:
Here -- for example -- you call the ExecuteLoaderCleanly, but I don't see how the old esp (now stored in eax) parameter could have been NULL.

Reply:
It’s not physically possible, I was trying to sanity check but you’re right this wouldn’t remotely happen

Reply:
ok then this should allow you to simplify further that execute.... function.